cmd/admin-handlers.go

```diff
@@ -694,10 +694,76 @@ func (adminAPI adminAPIHandlers) GetConfigHandler(w http.ResponseWriter, r *http
 	// returns local config.json.
 	configBytes, err := getPeerConfig(globalAdminPeers)
 	if err != nil {
-		writeErrorResponse(w, toAPIErrorCode(err), r.URL)
+		var adminErrCode APIErrorCode
+		if err == errXLWriteQuorum {
+			adminErrCode = ErrAdminConfigNoQuorum
+		} else {
+			adminErrCode = toAPIErrorCode(err)
+		}
+
+		writeErrorResponse(w, adminErrCode, r.URL)
 		errorIf(err, "Failed to get config from peers")
 		return
 	}
 
 	writeSuccessResponseJSON(w, configBytes)
 }
+
+// toAdminAPIErrCode - converts errXLWriteQuorum error to admin API
+// specific error.
+func toAdminAPIErrCode(err error) APIErrorCode {
+	switch err {
+	case errXLWriteQuorum:
+		return ErrAdminConfigNoQuorum
+	}
+	return toAPIErrorCode(err)
+}
+
+// SetConfigHandler - PUT /?config
+// - x-minio-operation = set
+func (adminAPI adminAPIHandlers) SetConfigHandler(w http.ResponseWriter, r *http.Request) {
+	// Get current object layer instance.
+	objectAPI := newObjectLayerFn()
+	if objectAPI == nil {
+		writeErrorResponse(w, ErrServerNotInitialized, r.URL)
+		return
+	}
+
+	// Validate request signature.
+	adminAPIErr := checkRequestAuthType(r, "", "", "")
+	if adminAPIErr != ErrNone {
+		writeErrorResponse(w, adminAPIErr, r.URL)
+		return
+	}
+
+	tmpFileName := "config-" + mustGetUUID()
+	n, err := writeTmpConfigPeers(globalAdminPeers, tmpFileName, r.Body)
+	if err != nil {
+		writeErrorResponse(w, toAdminAPIErrCode(err), r.URL)
+		return
+	}
+
+	// Take a lock on minio/config.json. NB minio is a reserved
+	// bucket name and wouldn't conflict with normal object
+	// operations.
+	configLock := globalNSMutex.NewNSLock(minioReservedBucket, "config.json")
+	configLock.Lock()
+	defer configLock.Unlock()
+
+	// Rename the temporary config file to config.json
+	err = commitConfigPeers(globalAdminPeers, tmpFileName)
+	if err != nil {
+		writeErrorResponse(w, toAdminAPIErrCode(err), r.URL)
+		return
+	}
+
+	// Send the length of config file contents. This is to be inline
+	// with any Write like operation. We return the number of
+	// bytes written.
+	contentLengthStr := strconv.FormatInt(int64(n), 10)
+	w.Header().Set("Content-Length", contentLengthStr)
+	writeSuccessResponseHeadersOnly(w)
+
+	// Restart all node for the modified config to take effect.
+	sendServiceCmd(globalAdminPeers, serviceRestart)
```

Comment:
The restart event could be racing with the response reaching the client.

The client should ideally have a good opportunity to know that the config update request is successful. If it gets an error response because the server is restarting the client may issue the request again.

Perhaps the restart request can be done in a defer with a 3 second delay? The delay is just to give the client a good chance to receive the response.